activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -208,7 +211,16 @@ def current_transaction
       end
 
       private
+
         NULL_TRANSACTION = NullTransaction.new
+
+        # Deallocate invalidated prepared statements outside of the transaction
+        def after_failure_actions(transaction, error)
+          return unless transaction.is_a?(RealTransaction)
+          return unless error.is_a?(ActiveRecord::PreparedStatementCacheExpired)
```

Comment:
if this is a Postgress specific error, I am not sure if we should have handled by all adapters.